### PR TITLE
cli: avoid logging during flag initialization

### LINKED ISSUE #30120: cli,log: spurious log file created for start --background
In some roachtest artifacts I see spurious log files created from the `cockroach start --background` process. For example:

```
I180911 14:27:07.082225 1 util/log/clog.go:1151  [config] file created at: 2018/09/11 14:27:07
I180911 14:27:07.082225 1 util/log/clog.go:1151  [config] running on machine: c688e08a1da4
I180911 14:27:07.082225 1 util/log/clog.go:1151  [config] binary: CockroachDB CCL v2.2.0-alpha.00000000-670-g83a7cb5 (x86_64-pc-linux-gnu, built 2018/09/11 14:25:15, go1.10.3)
I180911 14:27:07.082225 1 util/log/clog.go:1151  [config] arguments: [/home/roach/local/1/cockroach start --insecure --store=path=/home/roach/local/1/data --log-dir=/home/roach/local/1/logs --background --cache=6% --max-sql-memory=6% --port=26257 --http-port=26258 --locality=region=local,zone=local --host=127.0.0.1]
```

Another log file is created moments later:

```
I180911 14:27:07.151923 1 util/log/clog.go:1151  [config] file created at: 2018/09/11 14:27:07
I180911 14:27:07.151923 1 util/log/clog.go:1151  [config] running on machine: c688e08a1da4
I180911 14:27:07.151923 1 util/log/clog.go:1151  [config] binary: CockroachDB CCL v2.2.0-alpha.00000000-670-g83a7cb5 (x86_64-pc-linux-gnu, built 2018/09/11 14:25:15, go1.10.3)
I180911 14:27:07.151923 1 util/log/clog.go:1151  [config] arguments: [/home/roach/local/1/cockroach start --insecure --store=path=/home/roach/local/1/data --log-dir=/home/roach/local/1/logs --cache=6% --max-sql-memory=6% --port=26257 --http-port=26258 --locality=region=local,zone=local --host=127.0.0.1]
```

Notice that the arguments are identical modulo the `--background` flag. This doesn't reproduce for me on Mac OS X. Will test on Linux shortly.

### PR DESCRIPTION
**Pull request by petermattis**
Avoid logging during flag initialization which happens before we've had
a chance to examine the --background flag. This was spuriously creating
a log file that was mostly empty.

Fixes #30120

Release note: None

### CODE REVIEW

**Comment by petermattis** [comment]
This patches over the proximate cause of spurious log files. It would be nice to have a more robust way of not logging if `--background` is specified.

**Review by knz — APPROVED** [review]
:lgtm: this is not a bad solution. 

I am a bit unsure about the root cause though. The log package should not be creating a log file until the `start` command has decided that a log file needs to be created. (i.e. until log.logDir is set) I am surprised the thing you found caused a log file to appear in the first place.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123#-:-LMJd29N9WHeWomf8Woi:b-nv1r3m)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by petermattis** [comment]
> I am a bit unsure about the root cause though. The log package should not be creating a log file until the start command has decided that a log file needs to be created. (i.e. until log.logDir is set) I am surprised the thing you found caused a log file to appear in the first place.

I'm happy to let you take this over and track down what is going on there.

**Comment by knz** [comment]
I'd actually be glad to however it would need to wait a week or two. If you think that's inconvenient I'm happy to merge as-is

**Comment by petermattis** [comment]
The log message is being generated when parsing the flag. Specifically, `bytesOrPercentageValue.Set` calls `memoryPercentResolver` which was logging a message. A log file was created, because cockroach was started with the `--log-dir` flag specified and this flag occurred on the command line before the flag which used `memoryPercentResolver`. Specifically, if you start cockroach with `--log-dir=foo --cache=25%` you encounter this problem. If you instead start cockroach with `--cache=25% --log-dir=foo` you do not. 

Seems like we want to have a layer of indirection for the `--log-dir` flag so that we parse it, but do not change the value in `util/log` until later in the startup process.

**Comment by knz** [comment]
gotcha. Nice investigation, I would not have thought about argument ordering.

**Comment by petermattis** [comment]
Added a second commit to this PR which removes the `--log-{dir,file}*` flags for all commands except for `start` and `demo`. It also changes the handling of the `--log-dir` flag in a way that would have avoided the problem fixed by the first commit. PTAL.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/cliflags/flags.go, line 804 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30123#-LMOd2LmBbN0fowrVJEI:-LMOd2LmBbN0fowrVJEJ:b-nii11g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b115f168a869999e98f3262448b141a2db1ceb5/pkg/cli/cliflags/flags.go#L804)):*
> ```Go
> If non-empty, write log files in this directory. If empty, write log files to
> the <store-dir>/logs where <store-dir> is the directory of the first on disk
> store.
> ```

@piyush-singh or @jseldess the wording here and on the other help texts could be improved. I noticed that the default log directory stuff doesn't seem to be mentioned on our docs page: https://www.cockroachlabs.com/docs/v2.0/start-a-node.html#defaults

**Comment by petermattis** [comment]
Ping. The second commit needs a review.

File: pkg/cli/flags.go [hunk 1]
```diff
@@ -235,13 +235,17 @@ func init() {
 			// is only overriding the default value for the pflag package (and what
 			// is visible in help text), not the stdlib flag value.
 			flag.DefValue = "NONE"
+		case logflags.LogDirName,
+			logflags.LogFileMaxSizeName,
+			logflags.LogFilesCombinedMaxSizeName,
+			logflags.LogFileVerbosityThresholdName:
+			// The --log-dir* and --log-file* flags are specified only for the
+			// `start` command.
```
Comment (jseldess) [inline pkg/cli/flags.go]:
```
`start` and `demo` commands.
```
?

**Review by jseldess — APPROVED** [review]
Sorry for the delay. LGTM, with a few nits.

File: pkg/cli/cliflags/flags.go [hunk 2]
```diff
@@ -780,4 +780,34 @@ in the history of the cluster.`,
 		Name:        "align",
 		Description: `Align the output.`,
 	}
+
+	LogDir = FlagInfo{
+		Name: "log-dir",
+		Description: `
+If non-empty, write log files in this directory. If empty, write log files to
+the <store-dir>/logs where <store-dir> is the directory of the first on disk
```
Comment (jseldess) [inline pkg/cli/cliflags/flags.go]: I'd remove `the` from `the <store-dir>/logs`.

Comment (petermattis) [reply pkg/cli/cliflags/flags.go]: Done.

File: pkg/cli/flags.go (hunk 1 above)
Comment (petermattis) [reply pkg/cli/flags.go]: Yep. Done.

**Review by petermattis — COMMENTED** [review]
Thanks, Jesse. @knz Anything you'd like to add? Or @bdarnell who I believe had an opinion about the logging flags.

**Review by bdarnell — APPROVED** [review]
> Reviewed 1 of 8 files at r3, 7 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by knz — APPROVED** [review]
hey this change is actually pretty cool. :lgtm_strong: 

> Reviewed 1 of 8 files at r3, 6 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123#-:-LMdV9px17XqpSYnpkU_:b9sdlrh)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Comment by petermattis** [comment]
bors r=knz,bdarnell,jseldess

@jseldess I assume you've got the docs update for this on your plate (if there are any).

**Comment by petermattis** [comment]
`acceptance/bank/node-restart` flake (which is on my plate to fix). Retrying.

bors r=knz,bdarnell,jseldess

**Comment by jseldess** [comment]
@petermattis, yes, once this makes it into a release, our release notes script will pick up the change based on your release note text. And then we'll update the docs.

**Pull request merged**